src/parser/parser_verification.js

```diff
@@ -0,0 +1,150 @@
+    return;
+  }
+
+  return;
+}
+
+/**
+ * Verify and trigger warnings if node required attributes are not set.
+ * @param  {Node} node - The node element.
+ * @param  {Function} emit - Emit function used to trigger Warning event.
+ */
+function verifyRequiredAttributes(node, emit) {
+  if (!requiredValues[node.nodeName]) {
+    return;
+  }
+  const requiredAttributes = requiredValues[node.nodeName].attributes;
+  for (const attributeKey in requiredAttributes) {
+    const attributeName = requiredAttributes[attributeKey];
+    const attribute = node.getAttribute(attributeName);
+    if (!attribute) {
+      emitMissingValueWarning(
+        { node, name: attributeName, type: 'attribute' },
+        emit
+      );
+    }
+  }
+}
+
+/**
+ * Verify and trigger warnings if node required sub element are not set.
+ * @param  {Node} node - The node element
+ * @param  {Boolean} isAdInline - True if node is contained in a inline
+ * @param  {Function} emit - Emit function used to trigger Warning event.
+ */
+function verifyRequiredSubElements(node, emit, isAdInline) {
+  const required = requiredValues[node.nodeName];
+  if (!required) {
+    return;
+  }
+  const requiredSubElements = required.subElements;
+
+  for (const subElementKey in requiredSubElements) {
+    const subElementName = requiredSubElements[subElementKey];
+    const subElement = parserUtils.childByName(node, subElementName);
+    if (!subElement) {
+      emitMissingValueWarning(
+        { node, name: subElementName, type: 'sub element' },
+        emit
+      );
+    }
+  }
+
+  // When InLine format is used some node require at least one resource (i.e <NonLinear>, <Companion>, or <Icon>)
+  if (!isAdInline || !required.inLineRessources) {
+    return;
+  }
+
+  let ressourceFound = false;
+  for (const ressourceKey in required.inLineRessources) {
+    const ressource = required.inLineRessources[ressourceKey];
+    if (parserUtils.childByName(node, ressource)) {
+      ressourceFound = true;
+      break;
+    }
+  }
+  if (!ressourceFound) {
+    emitMissingValueWarning(
+      {
+        node,
+        name: `${required.inLineRessources.join(' or ')}`,
+        type: 'sub element'
+      },
+      emit
+    );
+  }
+}
+
+/**
+ * Check if a node is sub elements.
+ * @param  {Node} node - The node element.
+ * @returns {Boolean}
+ */
+function hasSubElements(node) {
+  if (
+    !node.children ||
+    node.children.length === 0 ||
+    (node.children.length === 1 &&
+      ['#text', '#cdata-section'].indexOf(node.children[0].nodeName) >= 0)
+  ) {
+    return false;
+  }
+  return true;
+}
+
+/**
+ * Trigger Warning for if a node is empty or has missing attributes/subelements
+ * @param  {Object} missingElement - Object containing the node with empty value, or with missing attribute/subElement name and type
+ * @param  {Function} emit - Emit function used to trigger Warning event.
+ * @emits  VastParser#VAST-warning
+ */
+function emitMissingValueWarning(missingElement, emit) {
```

Comment:
you could already destructure `missingElement`, that way we know what this function expects

Reply:
also maybe instead of having `{ node, name, type }` , you could pass `{ name, parentName, attributes, subElements }`. And in the message check if attributes is passed, or if subElements is passed, or neither, and make a message accordingly.